### PR TITLE
Implemented: logic to link group to product stores (#210)

### LINKED ISSUE #210: Link group to product store
## What is the motivation for adding/enhancing this feature?

Add option link a facility group to product stores, similar to how facilities are linked to groups. Clicking on chip directly opens modal with product stores with check boxes. Clicking on save from modal updates assoc.

Product stores
Overline
Product Store Name
note
Item
Select
Product Store ID
Item
Secondary text 2
Item

### PR DESCRIPTION
**Pull request by amansinghbais**
### Related Issues

Closes #210

### Short Description and Why It's Useful

Added support to link / unlink product stores to facility group.

### Screenshots of Visual Changes before/after (If There Are Any)

Screenshot from 2024-03-05 15-31-00
Screenshot from 2024-03-05 15-31-20

### Contribution and Currently Important Rules Acceptance

- [x] I read and followed [contribution rules](https://github.com/hotwax/facilities#contribution-guideline)

### CODE REVIEW

File: src/store/modules/facility/actions.ts [hunk 1]
```diff
@@ -744,10 +744,15 @@ const actions: ActionTree<FacilityState, RootState> = {
     stateGroups = stateGroups.filter((group: any) => !facilityGroupIds.includes(group.facilityGroupId))
 
     try {
-      const facilityCountByGroup = await FacilityService.fetchFacilityCountByGroup(facilityGroupIds)
+      const facilityCountByGroup = await FacilityService.fetchFacilityCountByGroup(JSON.parse(JSON.stringify(facilityGroupIds)))
```
Comment (ymaheshwari1) [inline src/store/modules/facility/actions.ts:747]: Why we have deep-cloned this value?

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/store/modules/facility/actions.ts (hunk 1 above)
Comment (amansinghbais) [reply src/store/modules/facility/actions.ts:747]: facilityGroupIds is list of ids which gets empty at the end of below api call. We again want's to use this facilityGroupIds in another api hence deep cloning it.

File: src/components/AddProductStoreToGroupModal.vue [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+          throw resp.data
+        }
+      } catch(err) {
+        logger.error(err)
+      }
+    },
+    isSelected(productStoreId: any) {
+      return this.selectedProductStoreValues.some((productStore: any) => productStore.productStoreId === productStoreId);
+    },
+    toggleProductStoreSelection(updatedStore: any) {
```
Comment (ymaheshwari1) [inline src/components/AddProductStoreToGroupModal.vue:121]:
```suggestion
    toggleProductStoreSelection(store: any) {
```

File: src/components/AddProductStoreToGroupModal.vue [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        }))
+      )
+
+      const hasFailedResponse = [...removeResponses, ...addResponses].some((response: any) => response.status === 'rejected')
+      if (hasFailedResponse) {
+        showToast(translate("Failed to associate some product stores to group."))
+      } else {
+        showToast(translate("Product stores associated to group successfully."))
+      }
+      this.fetchGroups()
```
Comment (ymaheshwari1) [inline src/components/AddProductStoreToGroupModal.vue:157]: Do not fetch all the groups again, just update the count for currentGroup in state.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

File: src/components/AddProductStoreToGroupModal.vue [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      this.fetchGroupsCount()
+      modalController.dismiss()
+    },
+    async fetchGroupsCount() {
+      const productStoreCountByGroup = await FacilityService.fetchProductStoreCountByGroup([this.group.facilityGroupId])
+      this.groups.map((group: any) => {
+        if(group.facilityGroupId === this.group.facilityGroupId) {
+          group.productStoreCount = productStoreCountByGroup[this.group.facilityGroupId]
+        }
+      })
```
Comment (ymaheshwari1) [inline src/components/AddProductStoreToGroupModal.vue:167]: Use find to update the group information, as when using map the iteration will not stop even when we found the facilityGroup and the iteration will go on for all the items in the groups array.

File: src/services/FacilityService.ts [hunk 5]
```diff
@@ -601,6 +602,58 @@ const fetchFacilityCountByGroup = async (facilityGroupIds: any): Promise<any> =>
   }, {})
 }
 
+const fetchProductStoreCountByGroup = async (facilityGroupIds: any): Promise<any> => {
```
Comment (ymaheshwari1) [inline src/services/FacilityService.ts:605]:
```suggestion
const fetchProductStoreCountByGroup = async (facilityGroupIds: Array<string>): Promise<any> => {
```

File: src/services/FacilityService.ts [hunk 6]
```diff
@@ -601,6 +602,58 @@ const fetchFacilityCountByGroup = async (facilityGroupIds: any): Promise<any> =>
...
+  // taking out the response from Promise.allSettled's 'value' field first 
+  const allResponseData = productStoreCountResponse.map((response: any) => response.value)
+    .reduce((responseData: any, response: any) => {
+      if (!hasError(response)) {
+        responseData.push(...response.data.docs)
+      }
+      return responseData
+    }, [])
+
+  return allResponseData.reduce((productStoreCountByGroup: any, responseData: any) => {
```
Comment (ymaheshwari1) [inline src/services/FacilityService.ts:647]: Handle the case when the count for some facilityGroup is not found or the api fails. in this case we need to display 0 as count on the UI.

**Review by ymaheshwari1 — CHANGES_REQUESTED** [review]

**Review by ymaheshwari1 — APPROVED** [review]

**Pull request merged**